openwisp_monitoring/views.py

```diff
@@ -32,8 +33,49 @@ def _get_additional_data(request, *args, **kwargs):
         """
         return {}
 
+    def _add_custom_date_group_map(self, time):
+        """
+        Adds custom date range time
+        grouping to chart.GROUP_MAP
+        """
+        # If custom date time grouping already exists
+        # with default time group then pop that time
+        if len(Chart.GROUP_MAP.items()) > 5:
+            Chart.GROUP_MAP.popitem()
+        group = '10m'
+        days = int(time.split('d')[0])
+        # custom grouping between 1 to 2 days
+        if days > 0 and days < 3:
+            group = '10m'
+        # custom grouping between 3 to 6 days (base 5)
+        elif days >= 3 and days < 7:
+            group = str(5 * round(((days / 3) * 20) / 5)) + 'm'
+        # custom grouping between 8 to 27 days
+        elif days > 7 and days < 28:
+            group = str(round(days / 7)) + 'h'
+        # custom grouping between 28 to 364 days
+        elif days >= 28 and days < 365:
+            group = str(round(days / 28)) + 'd'
+        Chart.GROUP_MAP.update({time: group})
+
+    def _validate_custom_date(self, start, end):
+        try:
+            datetime.strptime(start, '%Y-%m-%d %H:%M:%S')
+            datetime.strptime(end, '%Y-%m-%d %H:%M:%S')
+        except ValueError:
+            raise ValidationError(
+                'Incorrect custom date format, should be YYYY-MM-DD H:M:S'
+            )
+
     def get(self, request, *args, **kwargs):
         time = request.query_params.get('time', Chart.DEFAULT_TIME)
+        start_date = request.query_params.get('start', None)
+        end_date = request.query_params.get('end', None)
+        # if custom dates are provided
+        # then validate it & update chart.GROUP_MAP
+        if start_date and end_date:
+            self._validate_custom_date(start_date, end_date)
```

Comment:
are we also validating that dates are not made up here in this view?
We must ensure that we do not allow passing dates into the future, or that the beginning date is not older the end date.

Reply:
I'll look into this :+1: